src/tools/metrics/recorder.rs

```diff
@@ -0,0 +1,214 @@
+//! Metrics recording types and utilities.
+
+use std::collections::HashMap;
+
+use histogram::Histogram;
+use metrics::Key;
+use metrics_util::{
+    debugging::{DebugValue, DebuggingRecorder, Snapshotter},
+    CompositeKey, MetricKind,
+};
+
+// This is a false positive, `CompositeKey` does not depend on any interior mutable
+// types for its hashing implementation. Therefore, it is safe to use in our context.
+#[allow(clippy::mutable_key_type)]
+pub struct Snapshot(HashMap<CompositeKey, MetricVal>);
+
+impl Snapshot {
+    pub fn get_counter(&self, metric: &'static str) -> u64 {
+        let key = CompositeKey::new(MetricKind::Counter, Key::from_name(metric));
+        if let MetricVal::Counter(val) = *self.0.get(&key).unwrap() {
+            val
+        } else {
+            0
+        }
+    }
+
+    pub fn get_gauge(&self, metric: &'static str) -> f64 {
+        let key = CompositeKey::new(MetricKind::Gauge, Key::from_name(metric));
+        if let MetricVal::Gauge(val) = *self.0.get(&key).unwrap() {
```

Comment:
why some methods use `*self` and others `self`?

Reply:
[`HashMap::get`](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.get) returns an `Option<&v>`, the `*` is dereferencing what is returned from the `unwrap` call here. I'd recommend reading [the chapter on this in the book](https://doc.rust-lang.org/book/ch15-02-deref.html).